### DISCUSSION TITLE
Vulnerability Detector - Not updating old clients

### DISCUSSION THREAD

**Discussion by marcwalker1979** [wazuh/wazuh#25731] [General]
Hello,
How long should it take for the VD to update once a device has been patched or upgraded?

I have an old 2012 R2 server that has recently been upgraded to Server 2022, yet its still showing the Server as 2012 in VD.

The Agent has been removed and readded, and is correctly being reported under Active Agents as Server 2022, yet in VD it shows only one device of the same name, yet still on 2012 R2.

How do I clear out old entries from VD?

I also have quite a few MacBook's showing running versions on Monterey, that I know for a fact have been upgraded to Sonoma over 2 weeks ago, yet still showing as old versions?

This is a standalone Wazuh 4.9 deployment, upgraded from 4.8

Any help would be appreciated, as currently I cant trust what the Vulnerability Detector is telling me.

Thank you.

**Comment by Damian-Mangold** [comment]
@tsarquis88 please take it.

**Comment by tsarquis88** [comment]
Hi @marcwalker1979.

I'm still investigating this. I'm trying to reproduce this behavior you had.
I have one question: How did you upgrade the Windows OS? Which method you've used?
TIA!

**Comment by marcwalker1979** [comment]
Hello @tsarquis88 

We did an in-place upgrade... firstly upto 2019, then to 2022.

However I dont think that is the problem. I have VD reporting on other agents that I know have CVE's that have been patched... its even reporting on Agents that have been deleted from the system, yet the VD database seems to think these agents are still active.

How can I delete old agents from the VD database because I am guessing these are now different agents:

![VD Wazuh Issue 1](https://github.com/user-attachments/assets/c8300252-a22f-4809-b0b6-398da73622a0)
![VD Wazuh Issue 2](https://github.com/user-attachments/assets/1a87f7da-d44f-43bd-941a-463617cbde39)

**Comment by tsarquis88** [comment]
So, regarding the upgrade thing I did a test simulating an upgrade (downgrade in fact) from Ubuntu Bionic to Ubuntu Trusty. The results were positive: Wazuh did realize the agent is now on a different OS and the vulnerabilities were re-scanned. I'm sharing some images of the process below:
- Agent added. OS: Bionic
![image1](https://github.com/user-attachments/assets/6cd75017-293f-48f5-b7bc-08f85869ceab)
![image2](https://github.com/user-attachments/assets/986052ed-35f2-471e-8d2a-0192a5207372)
- Agent "replaced" with a new one. OS: Trusty
![image3](https://github.com/user-attachments/assets/b9df80ae-717c-4f84-ba3b-8f6e8732c674)
![image4](https://github.com/user-attachments/assets/a4c5cdb6-a32d-4543-8f03-0135324382c0)

In the last two images we can see that the agent OS is correctly detected, and the hostname and IP address are the same as the "first" agent.
Also, the vulnerabilities analysis changed completely after the OS changement. :green_circle: 

---

Now, if the vulnerabilities are indeed from a deleted agent, the scenario is different. We have seen this in the past.
I'm asking just in case: Did you tried restarting Wazuh Manager?

If there was a communication problem between Wazuh Manager and Wazuh Indexer, it is possible that the old information couldn't be correctly removed. If this is the case, you should see some warning/error messages in your `ossec.log` file. Please, take a look at it and tell me if you find anything. If you want to, you can share it to me (try to avoid sharing any sensitive information though).

**Comment by marcwalker1979** [comment]
Yes, the server has been rebooted multiple times, and wazuh-manager, indexer and dashboard restarted also.

There are indeed some errors in the log, not sure if these are related:

2024/09/18 12:57:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 12:57:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 12:58:08 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:299 at run(): DEBUG: Event type: 11 processed
2024/09/18 12:58:09 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:299 at run(): DEBUG: Event type: 11 processed
2024/09/18 12:58:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 12:58:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 12:59:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 12:59:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:00:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:00:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:01:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:01:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:02:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:02:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:03:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:03:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:04:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:04:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:05:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:05:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:06:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:06:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.
2024/09/18 13:07:49 wazuh-modulesd:vulnerability-scanner[1153] scanAgentList.hpp:247 at handleRequest(): DEBUG: Error executing query to fetch agent data for agents. Reason: DB query not synced.
2024/09/18 13:07:49 wazuh-modulesd:vulnerability-scanner[1153] scanOrchestrator.hpp:143 at operator()(): DEBUG: AgentReScanListException. Reason: Error executing rescan for multiple agents.

**Comment by tsarquis88** [comment]
Thanks for the logs @marcwalker1979.

I'm discussing this with the vulnerability detection team. Hope to have news soon.

**Comment by marcwalker1979** [comment]
Morning,

Not sure if the below helps.... its not looking too healthy to me:


2024/09/19 06:25:41 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '023' sync omitted due to abuse control.
2024/09/19 06:26:34 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '028' with the indexer.
2024/09/19 06:26:34 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '028' with the indexer.
2024/09/19 06:26:34 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:26:34 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '028' sync omitted due to abuse control.
2024/09/19 06:29:16 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '027' with the indexer.
2024/09/19 06:29:16 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '027' with the indexer.
2024/09/19 06:29:16 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:29:17 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '027' sync omitted due to abuse control.
2024/09/19 06:36:11 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '063' with the indexer.
2024/09/19 06:36:11 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '063' with the indexer.
2024/09/19 06:36:11 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:36:11 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '063' sync omitted due to abuse control.
2024/09/19 06:40:07 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '106' with the indexer.
2024/09/19 06:40:07 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '106' with the indexer.
2024/09/19 06:40:07 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:40:07 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '106' sync omitted due to abuse control.
2024/09/19 06:48:53 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '093' with the indexer.
2024/09/19 06:48:53 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '093' with the indexer.
2024/09/19 06:48:53 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:48:54 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '093' sync omitted due to abuse control.
2024/09/19 06:50:37 indexer-connector[1153] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '029' with the indexer.
2024/09/19 06:50:37 indexer-connector[1153] indexerConnector.cpp:446 at operator()(): WARNING: Failed to sync agent '029' with the indexer.
2024/09/19 06:50:37 indexer-connector[1153] indexerConnector.cpp:447 at operator()(): DEBUG: Error: No available server
2024/09/19 06:50:38 indexer-connector[1153] indexerConnector.cpp:129 at abuseControl(): DEBUG: Agent '029' sync omitted due to abuse control.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:02:10 indexer-connector[1153] indexerConnector.cpp:372 at operator()(): DEBUG: IndexerConnector is stopping, event processing will be skipped.
2024/09/19 07:04:05 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 2 seconds.
2024/09/19 07:04:05 indexer-connector[673] indexerConnector.cpp:482 at operator()(): WARNING: IndexerConnector initialization failed for index 'wazuh-states-vulnerabilities-wazuh-server', retrying until the connection is successful.
2024/09/19 07:04:05 indexer-connector[673] indexerConnector.cpp:366 at operator()(): DEBUG: Waiting for initialization thread to process events.
2024/09/19 07:04:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 4 seconds.
2024/09/19 07:04:11 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 8 seconds.
2024/09/19 07:04:19 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 16 seconds.
2024/09/19 07:04:35 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 32 seconds.
2024/09/19 07:05:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.
2024/09/19 07:06:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.
2024/09/19 07:07:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.
2024/09/19 07:08:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.
2024/09/19 07:09:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.
2024/09/19 07:10:07 indexer-connector[673] indexerConnector.cpp:474 at operator()(): DEBUG: Unable to initialize IndexerConnector for index 'wazuh-states-vulnerabilities-wazuh-server': No available server. Retrying in 60 seconds.

**Comment by tsarquis88** [comment]
Hi @marcwalker1979. Indeed, you have a problem with the communication between the Wazuh Indexer and the Wazuh Manager. This is one of the possible causes of the old data still being shown.

The `Error: No available server` message is telling that there is no indexer available.

Is your Wazuh Indexer up and running? 
What's the output of `filebeat test output`?
How is the indexer's health?
- `curl -X GET "http://<INDEXER_IP_ADDRESS>:9200/_cat/health?v"`

If the status is `yellow`, the manager won't consider that node as reliable and thus won't index the data.

**Comment by marcwalker1979** [comment]
Hi Tomas,

The cluster is in a yellow state:

{"cluster_name":"wazuh-cluster","status":"yellow","timed_out":false,"number_of_nodes":1,"number_of_data_nodes":1,"discovered_master":true,"discovered_cluster_manager":true,"active_primary_shards":151,"active_shards":151,"relocating_shards":0,"initializing_shards":0,"unassigned_shards":1,"delayed_unassigned_shards":0,"number_of_pending_tasks":0,"number_of_in_flight_fetch":0,"task_max_waiting_in_queue_millis":0,"active_shards_percent_as_number":99.3421052631579}

I have spun up a new instance and migrated all agents over, but happy to continue debugging the old server if you think its worth it?

Many thanks for your help.

**Comment by tsarquis88** [comment]
Hi @marcwalker1979 

The synchronization with the indexer won't when the state is yellow. So that's way there was missing/old information in the dashboard.

I believe that from v4.9.1, the sync in yellow state will be allowed (see https://github.com/wazuh/wazuh/issues/25416).

Is now your setup detecting/indexing correctly?

**Comment by marcwalker1979** [comment]
Hi Tomas,

OK but now sure why its in Yellow, or where to start troubleshooting how to get it back to Green. If you would like me to provide logs or continue to debug for future reference/help others then please let me know how I can help.

As mentioned, I have create a fresh instance and migrated all our agents over so the old server is no longer processing live traffic.

Kind regards,
Marc

**Comment by tsarquis88** [comment]
From [OpenSearch](https://opensearch.org/docs/1.1/opensearch/rest-api/cluster-health/) site:
> A yellow status means all primary shards are allocated to nodes, but some replicas aren’t.

Here are some blogs that explain how to attack this issue:
- [OpenSearch Yellow Status](https://opster.com/guides/opensearch/opensearch-basics/yellow-status/)
- [Why is my Amazon OpenSearch Service cluster in a red or yellow status?](https://repost.aws/knowledge-center/opensearch-red-yellow-status)

Another common reason behind this is when there is low disk space.

Hope it helps.

**Comment by Damian-Mangold** [comment]
I closed it because it had been inactive for more than 7 days. Please reopen it if necessary.
